File: chromadb/utils/embedding_functions/ollama_embedding_function.py
from chromadb.api.types import Embeddings, Documents, EmbeddingFunction, Space
from chromadb.utils.embedding_functions.schemas import validate_config_schema
from typing import List, Dict, Any
import numpy as np
from urllib.parse import urlparse

DEFAULT_MODEL_NAME = "chroma/all-minilm-l6-v2-f32"


class OllamaEmbeddingFunction(EmbeddingFunction[Documents]):
    """
    This class is used to generate embeddings for a list of texts using the Ollama Embedding API
    (https://github.com/ollama/ollama/blob/main/docs/api.md#generate-embeddings).
    """

    def __init__(
        self,
        url: str = "http://localhost:11434",
        model_name: str = DEFAULT_MODEL_NAME,
        timeout: int = 60,
    ) -> None:
        """
        Initialize the Ollama Embedding Function.

        Args:
            url (str): The Base URL of the Ollama Server (default: "http://localhost:11434").
            model_name (str): The name of the model to use for text embeddings.
                Defaults to "chroma/all-minilm-l6-v2-f32", for available models see https://ollama.com/library.
            timeout (int): The timeout for the API call in seconds. Defaults to 60.
        """
        try:
            from ollama import Client
        except ImportError:
            raise ValueError(
                "The ollama python package is not installed. Please install it with `pip install ollama`"
            )

        self.url = url
        self.model_name = model_name
        self.timeout = timeout

        # Adding this for backwards compatibility with the old version of the EF
        self._base_url = url
        if self._base_url.endswith("/api/embeddings"):
            parsed_url = urlparse(url)
            self._base_url = f"{parsed_url.scheme}://{parsed_url.netloc}"

        self._client = Client(host=self._base_url, timeout=timeout)

    def __call__(self, input: Documents) -> Embeddings:
        """
        Get the embeddings for a list of texts.

        Args:
            input (Documents): A list of texts to get embeddings for.

        Returns:
            Embeddings: The embeddings for the texts.

        Example:
            >>> ollama_ef = OllamaEmbeddingFunction()
            >>> texts = ["Hello, world!", "How are you?"]
            >>> embeddings = ollama_ef(texts)
        """
        # Call Ollama client
        response = self._client.embed(model=self.model_name, input=input)

        # Convert to numpy arrays
        return [
            np.array(embedding, dtype=np.float32)
            for embedding in response["embeddings"]
        ]

    @staticmethod
    def name() -> str:
        return "ollama"

    def default_space(self) -> Space:
        return "cosine"

    def supported_spaces(self) -> List[Space]:
        return ["cosine", "l2", "ip"]

    @staticmethod
    def build_from_config(config: Dict[str, Any]) -> "EmbeddingFunction[Documents]":
        url = config.get("url")
        model_name = config.get("model_name")
        timeout = config.get("timeout")

        if url is None or model_name is None or timeout is None:
            assert False, "This code should not be reached"

        return OllamaEmbeddingFunction(url=url, model_name=model_name, timeout=timeout)

    def get_config(self) -> Dict[str, Any]:
        return {"url": self.url, "model_name": self.model_name, "timeout": self.timeout}

    def validate_config_update(
        self, old_config: Dict[str, Any], new_config: Dict[str, Any]
    ) -> None:
        if "model_name" in new_config:
            raise ValueError(
                "The model name cannot be changed after the embedding function has been initialized."
            )

    @staticmethod
    def validate_config(config: Dict[str, Any]) -> None:
        """
        Validate the configuration using the JSON schema.

        Args:
            config: Configuration to validate

        Raises:
            ValidationError: If the configuration does not match the schema
        """
        validate_config_schema(config, "ollama")

File: clients/js/packages/chromadb-core/src/embeddings/OllamaEmbeddingFunction.ts
import type { IEmbeddingFunction } from "./IEmbeddingFunction";
import { validateConfigSchema } from "../schemas/schemaUtils";
type StoredConfig = {
  url: string;
  model_name: string;
};

export class OllamaEmbeddingFunction implements IEmbeddingFunction {
  name = "ollama";

  private readonly url: string;
  private readonly model: string;
  private ollamaClient?: any;

  constructor({
    url = "http://localhost:11434",
    model = "chroma/all-minilm-l6-v2-f32",
  }: { url?: string; model?: string } = {}) {
    // we used to construct the client here, but we need to async import the types
    // for the openai npm package, and the constructor can not be async
    if (url && url.endsWith("/api/embeddings")) {
      this.url = url.slice(0, -"/api/embeddings".length);
    } else {
      this.url = url;
    }
    this.model = model;
  }

  private async initClient() {
    if (this.ollamaClient) return;
    try {
      // @ts-ignore
      const { ollama } = await OllamaEmbeddingFunction.import();
      this.ollamaClient = new ollama.Ollama({ host: this.url });
    } catch (e) {
      // @ts-ignore
      if (e.code === "MODULE_NOT_FOUND") {
        throw new Error(
          "Please install the ollama package to use the OllamaEmbeddingFunction, `npm install -S ollama`",
        );
      }
      throw e;
    }
  }

  /** @ignore */
  static async import(): Promise<{
    // @ts-ignore
    ollama: typeof import("ollama");
  }> {
    try {
      // @ts-ignore
      const { ollama } = await import("ollama").then((m) => ({ ollama: m }));
      // @ts-ignore
      return { ollama };
    } catch (e) {
      throw new Error(
        "Please install Ollama as a dependency with, e.g. `npm install ollama`",
      );
    }
  }

  public async generate(texts: string[]) {
    await this.initClient();
    return await this.ollamaClient!.embed({
      model: this.model,
      input: texts,
    }).then((response: any) => {
      return response.embeddings;
    });
  }

  buildFromConfig(config: StoredConfig): OllamaEmbeddingFunction {
    return new OllamaEmbeddingFunction({
      url: config.url,
      model: config.model_name,
    });
  }

  getConfig(): StoredConfig {
    return {
      url: this.url,
      model_name: this.model,
    };
  }

  validateConfig(config: StoredConfig): void {
    validateConfigSchema(config, "ollama");
  }
}
